### ISSUE TITLE
Disable prepared statements for queries using partial indexes (index won't be used otherwise)

### ISSUE THREAD

**Issue by maia** [rails/rails#22566]
**Update**, see first comment: prepared statements fail to use partial indexes. Rails should warn when both are used in one app, or better: (manually or automatically) disable the use of prepared statements for queries that might include a partial index.

----

I'm seeing the following non-reproducible behavior on production and development (both Ruby 2.2.2, Rails 4.2.5. Postgres is 9.4.4 on production (heroku), 9.4.1 locally on OSX 10.11.2):

Rails/Postgres (whoever the culprit is) spends many seconds wasting realtime (not cputime or walltime) on a very specific query. If changing the query only slightly so it doesn't match a previously prepared statement (or by completely omitting prepared statements by skipping binding params), it will run fine:

````ruby
Benchmark.bm do |x|
  x.report { Tweet.where(recent: true).pluck('MIN("tweets"."tweeted_at")') }
  x.report { Tweet.where(recent: true).minimum(:tweeted_at) }
  x.report { Tweet.where(recent: true).pluck('MIN(tweeted_at)') }
  x.report { Tweet.where('recent = true').minimum(:tweeted_at) }
  x.report { Tweet.where('recent = true').pluck('MIN("tweets"."tweeted_at")') }
end
````
````SQL
(22300.0ms)  SELECT MIN("tweets"."tweeted_at") FROM "tweets" WHERE "tweets"."recent" = $1  [["recent", "t"]]
(21297.4ms)  SELECT MIN("tweets"."tweeted_at") FROM "tweets" WHERE "tweets"."recent" = $1  [["recent", "t"]]
(25.6ms)  SELECT MIN(tweeted_at) FROM "tweets" WHERE "tweets"."recent" = $1  [["recent", "t"]]
(1.2ms)  SELECT MIN("tweets"."tweeted_at") FROM "tweets" WHERE (recent = true)
(1.3ms)  SELECT MIN("tweets"."tweeted_at") FROM "tweets" WHERE (recent = true)
````
````
       user     system      total        real
  0.000000   0.020000   0.020000 ( 22.324030)
  0.010000   0.000000   0.010000 ( 21.301374)
  0.000000   0.010000   0.010000 (  0.028042)
  0.000000   0.000000   0.000000 (  0.003321)
  0.000000   0.000000   0.000000 (  0.002948)
````
Here the content of pg_prepared_statements (I'm omitting other queries that are not related to this case):
````
> q = "select name, statement, parameter_types from pg_prepared_statements WHERE length(statement) < 100"
> ActiveRecord::Base.connection.select_rows(q).each {|a| puts a.join(' : ') }
a13 : SELECT MIN(tweets.tweeted_at) FROM "tweets" WHERE "tweets"."recent" = $1 : {boolean}
a14 : SELECT MIN(tweeted_at) FROM "tweets" WHERE "tweets"."recent" = $1 : {boolean}
a1 : SELECT "tweets"."text" FROM "tweets" WHERE "tweets"."recent" = $1 AND "tweets"."language_id" = $2 : {boolean,integer}
a4 : SELECT MIN("tweets"."tweeted_at") FROM "tweets" WHERE "tweets"."recent" = $1 : {boolean}
a8 : SELECT MAX("tweets"."tweeted_at") FROM "tweets" WHERE "tweets"."recent" = $1 : {boolean}
a12 : SELECT min(tweets.tweeted_at) FROM "tweets" WHERE "tweets"."recent" = $1 : {boolean}

> ActiveRecord::Base.connection.select_value('select count(*) from pg_prepared_statements')
 => "14"
````
Here's what stackprof reports:
````
> StackProf.run(mode: :cpu, out: 'tmp/query-cpu.dump') { Tweet.where(recent: true).pluck('MIN("tweets"."tweeted_at")') }
$ stackprof tmp/query-cpu.dump --text --limit 5
==================================
  Mode: cpu(1000)
  Samples: 14 (0.00% miss rate)
  GC: 0 (0.00%)
==================================
     TOTAL    (pct)     SAMPLES    (pct)     FRAME
         2  (14.3%)           2  (14.3%)     Arel::Visitors::Visitor#dispatch
         1   (7.1%)           1   (7.1%)     ActiveRecord::Relation#initialize
         1   (7.1%)           1   (7.1%)     ThreadSafe::NonConcurrentCacheBackend#get_or_default
         1   (7.1%)           1   (7.1%)     ActiveRecord::SpawnMethods#merge
         1   (7.1%)           1   (7.1%)     ActiveRecord::Type::Value#binary?

> StackProf.run(mode: :wall, out: 'tmp/query-wall.dump') { Tweet.where(recent: true).pluck('MIN("tweets"."tweeted_at")') }
$ stackprof tmp/query-wall.dump --text --limit 5
==================================
  Mode: wall(1000)
  Samples: 2 (99.99% miss rate)
  GC: 0 (0.00%)
==================================
     TOTAL    (pct)     SAMPLES    (pct)     FRAME
         1  (50.0%)           1  (50.0%)     block in ActiveRecord::ConnectionAdapters::PostgreSQLAdapter#exec_cache
         1  (50.0%)           1  (50.0%)     block in Logger::LogDevice#write
         2 (100.0%)           0   (0.0%)     block in #<Module:0x007fa5c54405d0>.start
         2 (100.0%)           0   (0.0%)     ActiveRecord::ConnectionAdapters::AbstractAdapter#log
         2 (100.0%)           0   (0.0%)     ActiveRecord::ConnectionAdapters::PostgreSQLAdapter#exec_cache

> StackProf.run(mode: :object, out: 'tmp/query-object.dump') { Tweet.where(recent: true).pluck('MIN("tweets"."tweeted_at")') }
$ stackprof tmp/query-object.dump --text --limit 5
==================================
  Mode: object(1)
  Samples: 286 (0.00% miss rate)
  GC: 0 (0.00%)
==================================
     TOTAL    (pct)     SAMPLES    (pct)     FRAME
        33  (11.5%)          29  (10.1%)     ActiveRecord::Type::DateTime#fallback_string_to_time
        20   (7.0%)          16   (5.6%)     ActiveRecord::ConnectionAdapters::PostgreSQL::Utils#extract_schema_qualified_name
        24   (8.4%)          15   (5.2%)     ActiveRecord::QueryMethods#create_binds
        29  (10.1%)          11   (3.8%)     ActiveRecord::LogSubscriber#sql
        14   (4.9%)           9   (3.1%)     block in ActiveRecord::ConnectionAdapters::PostgreSQL::DatabaseStatements#exec_query
````
Any ideas on a possible reason?

**Comment by maia** [comment]
Addendum: It seems to be Postgres failing to use a partial index created by ActiveRecord on prepared statements.

See: schema.rb
````ruby
add_index "tweets", ["tweeted_at"], name: "index_tweets_on_recent_tweeted_at", where: "(recent = true)", using: :btree
add_index "tweets", ["tweeted_at"], name: "index_tweets_on_tweeted_at", using: :btree
````
(I have a total of 2.2M tweets, 38k out of these have recent=true, so the partial index is MUCH smaller)

The non-prepared statement properly uses the partial index:
````SQL
# EXPLAIN (ANALYZE, BUFFERS) SELECT MIN("tweets"."tweeted_at") FROM "tweets" WHERE "tweets"."recent" = true;
                                                                            QUERY PLAN
------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Result  (cost=0.88..0.89 rows=1 width=0) (actual time=23.665..23.666 rows=1 loops=1)
   Buffers: shared hit=1 read=16
   InitPlan 1 (returns $0)
     ->  Limit  (cost=0.29..0.88 rows=1 width=8) (actual time=23.658..23.659 rows=1 loops=1)
           Buffers: shared hit=1 read=16
           ->  Index Scan using index_tweets_on_recent_tweeted_at on tweets  (cost=0.29..22859.12 rows=38779 width=8) (actual time=23.656..23.656 rows=1 loops=1)
                 Index Cond: (tweeted_at IS NOT NULL)
                 Buffers: shared hit=1 read=16
 Planning time: 1.317 ms
 Execution time: 23.794 ms
(10 rows)
````

Now let's try the prepared statement:
````SQL
> ActiveRecord::Base.connection.select_rows('explain (analyze, buffers) EXECUTE a4(true)').each {|row| puts row}
Result  (cost=0.62..0.63 rows=1 width=0) (actual time=20662.080..20662.081 rows=1 loops=1)
  Buffers: shared hit=1768200 read=184748
  InitPlan 1 (returns $0)
    ->  Limit  (cost=0.43..0.62 rows=1 width=8) (actual time=20662.075..20662.075 rows=1 loops=1)
          Buffers: shared hit=1768200 read=184748
          ->  Index Scan using index_tweets_on_tweeted_at on tweets  (cost=0.43..223928.64 rows=1163381 width=8) (actual time=20662.073..20662.073 rows=1 loops=1)
                Index Cond: (tweeted_at IS NOT NULL)
                Filter: (recent = $1)
                Rows Removed by Filter: 2204631
                Buffers: shared hit=1768200 read=184748
Execution time: 20662.202 ms
````
Whoa! Postgres fails to use the partial index.

Is there anything ActiveRecord could do differently to create partial indexes, so Postgres properly uses them? What's the band-aid solution for this problem, I suspect I will have to disable prepared statements completely?

**Comment by maia** [comment]
Dear @tenderlove @sgrif @rafaelfranca , do you have any thoughts on this issue? Would adding a warning note in the documentation plus a method that forces skipping prepared statements for a specific query be the best solution? Or (doubtful) could AR identify which queries might use a partial index and automatically disable using prepared statements for these queries? Thanks!

**Comment by rafaelfranca** [comment]
@maia I need to check the issue closer but we already have a method to disable prepared statements for a specific query. http://api.rubyonrails.org/classes/ActiveRecord/ConnectionAdapters/AbstractAdapter.html#method-i-unprepared_statement. Could you check if with that you get better results?

**Comment by maia** [comment]
@rafaelfranca , I've completely missed the availability of `unprepared_statement`, thanks a lot, this will help now I know which queries I need to watch out for (basically all with a boolean that make use of a partial index in one of the boolean cases). 

Besides that, the Postgres query planner makes it difficult to reproduce that issue, but Tom Lane (a member of the core PostgreSQL development team) [mentioned](http://www.postgresql.org/message-id/1449.1359964764@sss.pgh.pa.us) it in 2013 in rare cases even in versions 9.2 and up:

> You might still have an issue if the partial index is only sometimes usable --- the choice logic might decide to go for the generic-plan approach anyway.  But if you've got a case where the optimal plan is all over the map like that, I wonder why you're using a prepared statement at all.

Now the question probably is if one can accept that in rare cases apps will cause extreme strain on the database without easy reproducability (I was bugged by the issue over many weeks until I finally realized that Postgres might use a different index in a prepared statement than in the unprepared statement), if a warning note in the documentation for partial indexes might help, or if there's a better solution. I managed to solve the issue, but there might be others...

**Comment by matthewd** [comment]
There's really nothing we can do here... and it seems outside our remit to even discuss it in our documentation: it's an implementation detail of certain versions of a particular RDBMS. (All current versions, perhaps, but I wouldn't count on that remaining so.) Note that our documentation doesn't even state what a partial index *is*; only how to create one using `add_index`.

... and [the PostgreSQL documentation](http://www.postgresql.org/docs/current/static/indexes-partial.html) already calls out this limitation:

> Matching takes place at query planning time, not at run time. As a result, parameterized query clauses do not work with a partial index. For example a prepared query with a parameter might specify "x < ?" which will never imply "x < 2" for all possible values of the parameter.

**Comment by maia** [comment]
@matthewd , thanks for your feedback. The last option that comes to my mind is to explicitly mention the option of `unprepared_statement`in the guides wherever prepared statements are explained – but as I cannot find the documentation of prepared statements in the guides at all (just the a link to tenderloves blog post in the 4.2 release notes), I'd have no suggestion on where to place that note. So I'll close this issue.

**Comment by pixeltrix** [comment]
Just been hit by this on [petition.parliament.uk][2] - turns out that after five `EXECUTE` calls for a prepared statement PostgreSQL will fallback to a generic query query if its cost estimate is less than the cost estimates of the specific query plans generated for the first five calls. Seems like the cost estimates for query plans using partial indexes in prepared statements are off - hoping that [this change in PostgreSQL 9.6][1] makes the estimates more accurate.

[1]: https://github.com/postgres/postgres/commit/f9aefcb91fc1f73fc43e384f660c120e515af931
[2]: https://petition.parliament.uk

**Comment by pixeltrix** [comment]
This doesn't seem to have been fixed in 9.6 so the way I fixed it was to make sure that PostgreSQL saw the literal true in the WHERE condition and not an argument to the prepared statement. See alphagov/e-petitions#636 for more details.
